presto-main/src/main/java/com/facebook/presto/execution/scheduler/CTEMaterializationTracker.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.scheduler;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/*
+ * Tracks the completion status of table-finish nodes that write temporary tables for CTE materialization.
+ * CTEMaterializationTracker manages a map of materialized CTEs and their associated materialization futures.
+ * When a stage includes a CTE table finish, it marks the corresponding CTE as materialized and completes
+ * the associated future. This signals the scheduler that some dependency has been resolved, prompting it
+ * to resume scheduling. The scheduler can call the clearAllFutures() method once it starts scheduling again
+ */
+public class CTEMaterializationTracker
+{
+    private final Map<String, SettableFuture<Void>> materializationFutures = new HashMap<>();
+
+    private final Map<String, Boolean> materializedCtes = new HashMap<>();
+
+    public synchronized ListenableFuture<Void> getFutureForCTE(String cteName)
+    {
+        return materializationFutures.compute(cteName, (key, existingFuture) -> {
+            if (existingFuture == null) {
+                return SettableFuture.create();
+            }
+            checkArgument(!existingFuture.isCancelled(), "CTE future was found in a cancelled state");
```

Comment:
this is still not thread safe, as the cancellation isn't atomic with clearing the futures, so it's possible to hit this condition (future cancelled, something tries to get the future before clearAllFutures is called).  I think it would be better to have the code consuming the futures gracefully handle receiving a cancelled future and not have to worry about synchronization.

Reply:
Part 1 `(future canceled, something tries to get the future before clearAllFutures is called)` can be fixed by clearing all before canceling. 

For part 2 `It's also possible for the future to be canceled after it's returned, but before the consumer checks it or even before it's returned, but after this argument check` 
I think this will never happen as blocked.cancel() is invoked in SqlQueryScheduler only after all stages are scheduled. At that point, all futures have already been returned to the SqlQueryScheduler, so the consumer will have completed scheduling (and return the blocked result) before this happens.


> I think it would be better to have the code consuming the futures gracefully handle receiving a canceled future and not have to worry about synchronization.

I think the code consuming the future can be other Schedulers and making them handle canceled futures might be challenging (not 100% sure if that is even needed yet)

Reply:
that's how it works today, but it's not really thread safe, as there's no guarantee that some other thread holding the future won't cancel it or that someone won't reorder the calling code without realizing the dependency and break the assumptions that clearAllFutures is always called before cancelling any future in the map. At the very least this needs some very clear code comments about the assumptions this class is making and requirements for anything using it.